### ISSUE TITLE
Error identifying Active Authentication Hash Algorithm - Chinese ePassport

### ISSUE THREAD

**Issue by rbrouwer** [AndyQ/NFCPassportReader#223]
The solution of #194 broke AA for recent Chinese passports.

Those will return 0 bytes when giving them an Expected Response Length of 65536. When undoing the changes to secure messaging and setting the Expected Response Length to 256. Those work just fine again.

I guess behaviour such as in JMRTD' AAAPDUSender is wanted; Guessing the correct length and even retrying with 65536 when 256 did not make sense.

**Comment by rbrouwer** [comment]
Setting Expected Response Length to 231 on [this line](https://github.com/AndyQ/NFCPassportReader/blob/32cc702d9f66cd8370ba1f6d2e59ad7eae284f4a/Sources/NFCPassportReader/TagReader.swift#L55) works for those documents as well. Guessing the extended format is not being liked.

**Comment by AndyQ** [comment]
Thats really annoying!  will have a re-think

**Comment by rbrouwer** [comment]
I do not believe that would work. Some documents (like that Australian passport in #194) will actually return OK with a partial signature (cutting off whatever did not fit). The implementation in aa_test would return that response and not go for an extended read.

This of course would work fine for this one document I now mentioned in this issue, but that other one would again break.

**Comment by AndyQ** [comment]
Hmmm thats a good point.  Wonder if it would work the other way round?  try first with extended read and if that failed then try standard?

**Comment by AndyQ** [comment]
Ah of course - small change! - currently doesn't check error - any better?

I'm really interested if the passport will allow the normal read after the extended read fails!

**Comment by rbrouwer** [comment]
Well... Apparently that will give you:
(Extended try)
sw - 0x67, sw2 - 0x00, and no data

and then 
(Normal try)
sw - 0x69, sw2 - 0x88, and no data

So apparently not.

**Comment by furkanisik000** [comment]
?

**Comment by baskurtbey** [comment]
@rbrouwer  Did you solve the problem

**Comment by baskurtbey** [comment]
The `doInternalAuthentication` function is actually not working. Active Authentication (AA) is not being created.

**Comment by AndyQ** [comment]
> Well... Apparently that will give you: (Extended try) sw - 0x67, sw2 - 0x00, and no data
> 
> and then (Normal try) sw - 0x69, sw2 - 0x88, and no data
> 
> So apparently not.

Thats a shame!  Not quite sure what to do here really.  I could add a flag to enable extended mode BUT you'd have to know that the passport supports/doesn't support that.  I'll did a little more into JMRTD code and see if I can see what they did to solve this.  I had a quick look thorugh and they had some functions in bouncy castle that could get certain info about the keys that I haven't yet figured out how to do (I'm sure its possible though).

**Comment by AndyQ** [comment]
> The `doInternalAuthentication` function is actually not working. Active Authentication (AA) is not being created.

Could you try with version 2.1.1 and see if that works for you (that doesn't enable extended read) and has been working for most passports (except newer Australian and probably other newer ones)?

**Comment by baskurtbey** [comment]
Hello, I tried but unfortunately there was an issue. It doesn't decode, most likely. The ID is a Turkish ID and it should normally be supported.

**Comment by baskurtbey** [comment]
It works seamlessly on Android using JMRTD.

**Comment by AndyQ** [comment]
So what I've done for the moment is instead of enabling extended mode for AA by default, we now have an additional flag set on PassportReader init - useExtendedMode.  Not ideal but will think more.

Also, as per #224 - We're now using OpenSSL 1.1.2300 which is a signed release and includes a privacy policy.

Currently main branch only,

**Comment by baskurtbey** [comment]
The latest update solved my problem, thank you. You saved me from a big issue; I had been researching it for a week.  🙏❤️

**Comment by danydev** [comment]
@AndyQ looks like it's working for @baskurtbey maybe it's time for a release? ;)

**Comment by AndyQ** [comment]
It's definitely getting closer. I'm still not 100% happy as I do t yet know what passport support extended reads and which don't!  It looks like some Chinese ones don't but I'm sure there are others too. So I can't yet say when you should use extended read or not. JMRTD does some best guesses based on key lengths BITBi haven't yet figured out how to replicate that in OpenSSL (JMRTD uses bouncycastle which has different APIs).

**Comment by tomgi** [comment]
@andyq for what it's worth, ICAO specifies in https://www.icao.int/publications/documents/9303_p11_cons_en.pdf that
> Note.— It should be noted that when using key lengths exceeding 1 848 bits (if Secure Messaging with 3DES is used) / 1 792 bits (if Secure Messaging with AES is used) in Active Authentication with Secure Messaging, Extended Length APDUs MUST be supported by the eMRTD chip and the Inspection System.

For Australian R-series EPassports, DG14 includes `id-PACE-ECDH-GM-AES-CBC-CMAC-256` ([0.4.0.127.0.7.2.2.4.2.4](https://oid-rep.orange-labs.fr/get/0.4.0.127.0.7.2.2.4.2.4)) as the only supported PACE algorithm. 
This is a 256-bit key, so based on that we know the chip MUST support extended length APDUs.

Logs from [here](https://github.com/AndyQ/NFCPassportReader/blob/89813bafed0490d1dab17e1c2c243e41bebcd455/Sources/NFCPassportReader/PACEHandler.swift#L94) are:
```
doPace - inpit parameters
paceOID - 0.4.0.127.0.7.2.2.4.2.4
parameterSpec - 933
mappingType - Generic Mapping
agreementAlg - ECDH
cipherAlg - AES
digestAlg - SHA-256
keyLength - 256
```

**Comment by AndyQ** [comment]
The problem here is that you need to know ni advance whether extended mode is supported or not.  Its not something you can try and if it fails switch to non-extended mode.  The NFC session gets aborted and you have to restart with the discovery again (which is a pain). I did spend some time looking at this but haven't yet figured out a way round this.
